File: mobile/shared/src/commonMain/kotlin/pro.aeternum/domain/model/Rosary.kt
package pro.aeternum.domain.model

import pro.aeternum.data.model.RosaryResponse

internal data class Rosary(
    val id: String,
    val title: String,
    val subtitle: String,
    val groups: List<Group>,
    val prayers: List<Prayer>,
)

internal data class Group(
    val steps: List<Step>,
)

internal data class Step(
    val type: String,
    val count: Int,
)

internal data class Prayer(
    val type: String,
    val title: String,
    val subtitle: String,
    val paragraphs: List<String>,
)

internal fun RosaryResponse.toDomainModel(): Rosary = Rosary(
    id = this.id,
    title = this.title,
    subtitle = this.subtitle,
    groups = this.groups.map { groupResponse ->
        Group(
            steps = groupResponse.steps.map { stepResponse ->
                Step(
                    type = stepResponse.type,
                    count = stepResponse.count,
                )
            }
        )
    },
    prayers = this.prayers.map { prayerResponse ->
        Prayer(
            type = prayerResponse.type,
            title = prayerResponse.title,
            subtitle = prayerResponse.subtitle,
            paragraphs = prayerResponse.paragraphs,
        )
    }
)

File: backend/pages/api/rosary.ts
import type { NextApiRequest, NextApiResponse } from 'next'
import { Group, Rosary } from "@/data/structure/Rosary"
import { prayerDataSource } from '@/data/source/PrayerDataSource';
import { rosaryDataSource } from '@/data/source/RosaryDataSource';
import { Prayer } from '@/data/structure/Prayer';

type RosaryResponse = {
    id: string
    title: string
    subtitle: string
    groups: Group[]
    prayers: Prayer[]
}

type Query = {
    id: string
}

export default function handler(req: NextApiRequest, res: NextApiResponse<RosaryResponse | Error>) {
    const { id } = req.query as Query

    try {
        const rosary: Rosary = rosaryDataSource.getSingleRosary(id)

        const response: RosaryResponse = {
            id: rosary.id,
            title: rosary.title,
            subtitle: rosary.subtitle,
            groups: rosary.groups,
            prayers: prayerDataSource.getPrayersForRosary(rosary)
        }

        res.status(200).json(response)
    } catch (exception) {
        res.status(401).json(exception as Error)
    }
}